File: internal/engine/engine.go
package engine

import (
	"context"
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"time"

	"github.com/cbaumont/nina/internal/llm"
	"github.com/cbaumont/nina/internal/runner"
	"github.com/cbaumont/nina/internal/state"
	"github.com/cbaumont/nina/internal/workspace"
)

type State string

const (
	StateIdle     State = "idle"
	StateScaffold State = "scaffold"
	StateDrive    State = "drive"
	StateDone     State = "done"
)

type EventKind string

const (
	EventTextDelta   EventKind = "text_delta"
	EventInfo        EventKind = "info"
	EventPlanSet     EventKind = "plan_set"
	EventStepStarted EventKind = "step_started"
	EventReview      EventKind = "review"
	EventSessionDone EventKind = "session_done"
	EventConfirm     EventKind = "confirm"
	EventCommandRun  EventKind = "command_run"
)

type Event struct {
	Kind    EventKind
	Text    string
	Step    int
	Plan    *Plan
	Verdict string
	Confirm *ConfirmRequest
}

// ConfirmRequest asks the user to approve a command the model proposed.
// The engine blocks until an answer is sent on Reply.
type ConfirmRequest struct {
	Command string
	Reason  string
	Reply   chan ConfirmAnswer
}

type ConfirmAnswer struct {
	Approve bool
	Always  bool
}

type Plan struct {
	Title string
	Steps []llm.PlanStep
}

const DialLevel = 1

const (
	commandTimeout   = 2 * time.Minute
	maxReadFileBytes = 32 * 1024
)

type Engine struct {
	client      llm.Client
	ws          *workspace.Workspace
	dir         string
	emit        func(Event)
	conv        *llm.Conversation
	sessionID   string
	goal        string
	state       State
	plan        Plan
	stepIndex   int
	snapshots   int
	lastRef     string
	review      *llm.SubmitReviewInput
	autoApprove map[string]bool
}

func New(client llm.Client, ws *workspace.Workspace, dir string, emit func(Event)) *Engine {
	return &Engine{
		client:      client,
		ws:          ws,
		dir:         dir,
		emit:        emit,
		sessionID:   time.Now().Format("20060102-150405"),
		state:       StateIdle,
		conv:        &llm.Conversation{System: systemPrompt()},
		autoApprove: map[string]bool{},
	}
}

func (e *Engine) State() State   { return e.state }
func (e *Engine) Plan() Plan     { return e.plan }
func (e *Engine) StepIndex() int { return e.stepIndex }
func (e *Engine) Goal() string   { return e.goal }

// Restore rebuilds the engine from a saved session so it can continue
// where it left off. Call before any other engine method.
func (e *Engine) Restore(sess *state.Session, messages []llm.Message) {
	e.sessionID = sess.SessionID
	e.goal = sess.Goal
	e.state = State(sess.State)
	e.plan = Plan{Title: sess.PlanTitle, Steps: sess.Steps}
	e.stepIndex = sess.StepIndex
	e.snapshots = sess.Snapshots
	e.lastRef = sess.LastRef
	e.conv.Messages = messages
}

// persist saves the session and transcript to .nina/; failures are
// reported to the user but never interrupt the session.
func (e *Engine) persist() {
	sess := &state.Session{
		SessionID: e.sessionID,
		Goal:      e.goal,
		State:     string(e.state),
		PlanTitle: e.plan.Title,
		Steps:     e.plan.Steps,
		StepIndex: e.stepIndex,
		Snapshots: e.snapshots,
		LastRef:   e.lastRef,
	}
	if err := state.Save(e.dir, sess, e.conv.Messages); err != nil {
		e.emit(Event{Kind: EventInfo, Text: "Warning: could not save session state: " + err.Error()})
	}
}

func (e *Engine) Start(ctx context.Context, goal string) error {
	if e.state != StateIdle {
		return fmt.Errorf("session already started")
	}
	e.state = StateScaffold
	e.goal = goal
	e.conv.AddUser(startPrompt(goal))
	if _, err := e.converseLoop(ctx); err != nil {
		e.state = StateIdle
		return err
	}
	if len(e.plan.Steps) == 0 {
		e.state = StateIdle
		return fmt.Errorf("the model did not produce a task plan; try rephrasing the goal")
	}
	if err := e.snapshot(); err != nil {
		return err
	}
	e.state = StateDrive
	e.persist()
	e.emit(Event{Kind: EventStepStarted, Step: e.stepIndex})
	return nil
}

func (e *Engine) Done(ctx context.Context) error {
	if e.state != StateDrive {
		return fmt.Errorf("no step in progress")
	}
	previousRef := e.lastRef
	if err := e.snapshot(); err != nil {
		return err
	}
	diff, err := e.ws.Diff(previousRef, e.lastRef)
	if err != nil {
		return err
	}
	if strings.TrimSpace(diff) == "" {
		e.emit(Event{Kind: EventInfo, Text: "No changes since the last snapshot — edit the files in your editor, then /done."})
		return nil
	}

	e.review = nil
	step := e.plan.Steps[e.stepIndex]
	e.conv.AddUser(reviewPrompt(step, diff))
	if _, err := e.converseLoop(ctx); err != nil {
		return err
	}
	if e.review == nil {
		e.emit(Event{Kind: EventInfo, Text: "The navigator did not submit a verdict; treating this step as still in progress."})
		return nil
	}
	e.emit(Event{Kind: EventReview, Verdict: e.review.Verdict, Text: e.review.Feedback, Step: e.stepIndex})
	if e.review.Verdict != "pass" {
		e.persist()
		return nil
	}

	e.stepIndex++
	if e.stepIndex >= len(e.plan.Steps) {
		e.state = StateDone
		e.persist()
		e.emit(Event{Kind: EventSessionDone})
		return nil
	}
	e.conv.AddUser(instructPrompt(e.stepIndex, e.plan.Steps[e.stepIndex]))
	if _, err := e.converseLoop(ctx); err != nil {
		return err
	}
	e.persist()
	e.emit(Event{Kind: EventStepStarted, Step: e.stepIndex})
	return nil
}

func (e *Engine) UserMessage(ctx context.Context, text string) error {
	if e.state == StateIdle {
		return fmt.Errorf("session not started")
	}
	e.conv.AddUser(text)
	_, err := e.converseLoop(ctx)
	if err == nil {
		e.persist()
	}
	return err
}

func (e *Engine) converseLoop(ctx context.Context) (llm.Turn, error) {
	for {
		turn, err := e.client.Converse(ctx, e.conv, func(text string) {
			e.emit(Event{Kind: EventTextDelta, Text: text})
		})
		if err != nil {
			return llm.Turn{}, err
		}
		if len(turn.ToolCalls) == 0 {
			return turn, nil
		}
		results := make([]llm.ToolResult, 0, len(turn.ToolCalls))
		for _, call := range turn.ToolCalls {
			results = append(results, e.execTool(ctx, call))
		}
		e.conv.AddToolResults(results)
	}
}

func (e *Engine) execTool(ctx context.Context, call llm.ToolCall) llm.ToolResult {
	switch call.Name {
	case llm.ToolWriteFile:
		return e.execWriteFile(call)
	case llm.ToolSetPlan:
		return e.execSetPlan(call)
	case llm.ToolSubmitReview:
		return e.execSubmitReview(call)
	case llm.ToolRunCommand:
		return e.execRunCommand(ctx, call)
	case llm.ToolReadFile:
		return e.execReadFile(call)
	default:
		return llm.ToolResult{ToolCallID: call.ID, Content: fmt.Sprintf("unknown tool %q", call.Name), IsError: true}
	}
}

func (e *Engine) execRunCommand(ctx context.Context, call llm.ToolCall) llm.ToolResult {
	var input llm.RunCommandInput
	if err := json.Unmarshal(call.Input, &input); err != nil {
		return llm.ToolResult{ToolCallID: call.ID, Content: "invalid run_command input: " + err.Error(), IsError: true}
	}
	if strings.TrimSpace(input.Command) == "" {
		return llm.ToolResult{ToolCallID: call.ID, Content: "run_command needs a non-empty command", IsError: true}
	}
	if !e.autoApprove[input.Command] {
		reply := make(chan ConfirmAnswer, 1)
		e.emit(Event{Kind: EventConfirm, Confirm: &ConfirmRequest{Command: input.Command, Reason: input.Reason, Reply: reply}})
		var answer ConfirmAnswer
		select {
		case answer = <-reply:
		case <-ctx.Done():
			return llm.ToolResult{ToolCallID: call.ID, Content: "cancelled", IsError: true}
		}
		if answer.Always && answer.Approve {
			e.autoApprove[input.Command] = true
		}
		if !answer.Approve {
			return llm.ToolResult{ToolCallID: call.ID, Content: "The learner declined to run this command. Continue without it, or propose an alternative and explain why it is needed."}
		}
	}
	result, err := runner.Run(ctx, e.dir, input.Command, commandTimeout)
	if err != nil {
		return llm.ToolResult{ToolCallID: call.ID, Content: err.Error(), IsError: true}
	}
	e.emit(Event{Kind: EventCommandRun, Text: commandOutputMarkdown(input.Command, result)})
	return llm.ToolResult{ToolCallID: call.ID, Content: commandToolContent(result)}
}

func commandOutputMarkdown(command string, result runner.Result) string {
	var b strings.Builder
	fmt.Fprintf(&b, "```console\n$ %s\n", command)
	if result.Output != "" {
		b.WriteString(result.Output + "\n")
	}
	b.WriteString("```\n")
	switch {
	case result.TimedOut:
		fmt.Fprintf(&b, "*timed out after %s and was killed*", commandTimeout)
	case result.ExitCode == 0:
		b.WriteString("*exit code 0*")
	default:
		fmt.Fprintf(&b, "*exit code %d*", result.ExitCode)
	}
	return b.String()
}

func commandToolContent(result runner.Result) string {
	content := fmt.Sprintf("exit code: %d", result.ExitCode)
	if result.TimedOut {
		content = fmt.Sprintf("command timed out after %s and was killed\n%s", commandTimeout, content)
	}
	if result.Output == "" {
		return content + "\n(no output)"
	}
	return content + "\n" + result.Output
}

func (e *Engine) execReadFile(call llm.ToolCall) llm.ToolResult {
	var input llm.ReadFileInput
	if err := json.Unmarshal(call.Input, &input); err != nil {
		return llm.ToolResult{ToolCallID: call.ID, Content: "invalid read_file input: " + err.Error(), IsError: true}
	}
	path, err := e.safePath(input.Path)
	if err != nil {
		return llm.ToolResult{ToolCallID: call.ID, Content: err.Error(), IsError: true}
	}
	data, err := os.ReadFile(path)
	if err != nil {
		return llm.ToolResult{ToolCallID: call.ID, Content: err.Error(), IsError: true}
	}
	if len(data) > maxReadFileBytes {
		return llm.ToolResult{ToolCallID: call.ID, Content: string(data[:maxReadFileBytes]) + "\n[file truncated]"}
	}
	return llm.ToolResult{ToolCallID: call.ID, Content: string(data)}
}

func (e *Engine) execWriteFile(call llm.ToolCall) llm.ToolResult {
	if e.state != StateScaffold {
		return llm.ToolResult{
			ToolCallID: call.ID,
			Content:    "Denied by the typing dial policy: at dial level 1 you may only write files while scaffolding the project. Instruct the learner to write this code themselves.",
			IsError:    true,
		}
	}
	var input llm.WriteFileInput
	if err := json.Unmarshal(call.Input, &input); err != nil {
		return llm.ToolResult{ToolCallID: call.ID, Content: "invalid write_file input: " + err.Error(), IsError: true}
	}
	path, err := e.safePath(input.Path)
	if err != nil {
		return llm.ToolResult{ToolCallID: call.ID, Content: err.Error(), IsError: true}
	}
	if err := os.MkdirAll(filepath.Dir(path), 0o755); err != nil {
		return llm.ToolResult{ToolCallID: call.ID, Content: err.Error(), IsError: true}
	}
	if err := os.WriteFile(path, []byte(input.Content), 0o644); err != nil {
		return llm.ToolResult{ToolCallID: call.ID, Content: err.Error(), IsError: true}
	}
	e.emit(Event{Kind: EventInfo, Text: "Nina wrote `" + input.Path + "`"})
	return llm.ToolResult{ToolCallID: call.ID, Content: "wrote " + input.Path}
}

func (e *Engine) execSetPlan(call llm.ToolCall) llm.ToolResult {
	var input llm.SetPlanInput
	if err := json.Unmarshal(call.Input, &input); err != nil {
		return llm.ToolResult{ToolCallID: call.ID, Content: "invalid set_plan input: " + err.Error(), IsError: true}
	}
	if len(input.Steps) == 0 {
		return llm.ToolResult{ToolCallID: call.ID, Content: "a plan needs at least one step", IsError: true}
	}
	e.plan = Plan{Title: input.Title, Steps: input.Steps}
	e.emit(Event{Kind: EventPlanSet, Plan: &e.plan})
	return llm.ToolResult{ToolCallID: call.ID, Content: fmt.Sprintf("plan set with %d steps", len(input.Steps))}
}

func (e *Engine) execSubmitReview(call llm.ToolCall) llm.ToolResult {
	var input llm.SubmitReviewInput
	if err := json.Unmarshal(call.Input, &input); err != nil {
		return llm.ToolResult{ToolCallID: call.ID, Content: "invalid submit_review input: " + err.Error(), IsError: true}
	}
	if input.Verdict != "pass" && input.Verdict != "retry" {
		return llm.ToolResult{ToolCallID: call.ID, Content: "verdict must be pass or retry", IsError: true}
	}
	e.review = &input
	return llm.ToolResult{ToolCallID: call.ID, Content: "verdict recorded"}
}

func (e *Engine) safePath(rel string) (string, error) {
	if rel == "" || filepath.IsAbs(rel) {
		return "", fmt.Errorf("path must be relative to the workspace: %q", rel)
	}
	root := filepath.Clean(e.dir)
	path := filepath.Join(root, rel)
	if path != root && !strings.HasPrefix(path, root+string(os.PathSeparator)) {
		return "", fmt.Errorf("path escapes the workspace: %q", rel)
	}
	return path, nil
}

func (e *Engine) snapshot() error {
	ref := workspace.SnapshotRef(e.sessionID, e.snapshots)
	if _, err := e.ws.Snapshot(ref); err != nil {
		return err
	}
	e.lastRef = ref
	e.snapshots++
	return nil
}

File: python/src/nina/engine.py
from __future__ import annotations

import asyncio
import time
from collections.abc import Callable
from pathlib import Path

from nina import profile as profile_module
from nina import prompts, runner, state
from nina.agent.session import AgentSession, RateLimited, TextDelta, ToolHandler, TurnComplete
from nina.events import (
    EVENT_COMMAND_RUN,
    EVENT_CONFIRM,
    EVENT_INFO,
    EVENT_PLAN_SET,
    EVENT_REVIEW,
    EVENT_SESSION_DONE,
    EVENT_STEP_STARTED,
    EVENT_TEXT_DELTA,
    STATE_DONE,
    STATE_DRIVE,
    STATE_IDLE,
    STATE_PROPOSE,
    STATE_SCAFFOLD,
    ConfirmAnswer,
    ConfirmRequest,
    Event,
    Plan,
)
from nina.profile import Profile
from nina.runner import Result
from nina.tools import (
    TOOL_READ_FILE,
    TOOL_RUN_COMMAND,
    TOOL_SET_PLAN,
    TOOL_SUBMIT_REVIEW,
    TOOL_UPDATE_PLAN,
    TOOL_WRITE_FILE,
    SubmitReviewInput,
    ToolResult,
    plan_steps_from,
)
from nina.workspace import Workspace, snapshot_ref

COMMAND_TIMEOUT = 120.0
MAX_READ_FILE_BYTES = 32 * 1024


class Engine:
    def __init__(
        self, ws: Workspace, dir: str, prof: Profile, emit: Callable[[Event], None]
    ) -> None:
        self.ws = ws
        self.dir = dir
        self.emit = emit
        self.session_id = time.strftime("%Y%m%d-%H%M%S")
        self.goal = ""
        self.state = STATE_IDLE
        self.plan = Plan()
        self.step_index = 0
        self.snapshots = 0
        self.last_ref = ""
        self.review: SubmitReviewInput | None = None
        self.profile = prof
        self.system_prompt = prompts.system_prompt(prof)
        self.session: AgentSession | None = None
        self._auto_approve: set[str] = set()

    def tool_handlers(self) -> dict[str, ToolHandler]:
        return {
            TOOL_WRITE_FILE: self._write_file,
            TOOL_SET_PLAN: self._set_plan,
            TOOL_UPDATE_PLAN: self._update_plan,
            TOOL_SUBMIT_REVIEW: self._submit_review,
            TOOL_RUN_COMMAND: self._run_command,
            TOOL_READ_FILE: self._read_file,
        }

    def update_profile(self, prof: Profile) -> None:
        self.profile = prof
        self.system_prompt = prompts.system_prompt(prof)
        profile_module.save(self.dir, prof)
        if self.state != STATE_IDLE:
            self.persist()

    def restore(self, sess: state.Session) -> None:
        self.session_id = sess.session_id
        self.goal = sess.goal
        self.state = sess.state
        self.plan = Plan(title=sess.plan_title, steps=list(sess.steps))
        self.step_index = sess.step_index
        self.snapshots = sess.snapshots
        self.last_ref = sess.last_ref

    def persist(self) -> None:
        sdk_session_id = self.session.session_id if self.session is not None else None
        sess = state.Session(
            session_id=self.session_id,
            goal=self.goal,
            state=self.state,
            plan_title=self.plan.title,
            steps=list(self.plan.steps),
            step_index=self.step_index,
            snapshots=self.snapshots,
            last_ref=self.last_ref,
            sdk_session_id=sdk_session_id,
        )
        try:
            state.save(self.dir, sess)
        except OSError as err:
            self.emit(Event(kind=EVENT_INFO, text=f"Warning: could not save session state: {err}"))

    async def start(self, goal: str) -> None:
        if self.state != STATE_IDLE:
            raise RuntimeError("session already started")
        self.state = STATE_PROPOSE
        self.goal = goal
        try:
            await self._converse(prompts.propose_prompt(goal))
        except Exception:
            self.state = STATE_IDLE
            raise
        self.persist()

    async def done(self) -> None:
        if self.state != STATE_DRIVE:
            raise RuntimeError("no step in progress")
        previous_ref = self.last_ref
        self._snapshot()
        diff = self.ws.diff(previous_ref, self.last_ref)
        if diff.strip() == "":
            self.emit(
                Event(
                    kind=EVENT_INFO,
                    text=(
                        "No changes since the last snapshot — edit the files in your "
                        "editor, then /done."
                    ),
                )
            )
            return

        self.review = None
        step = self.plan.steps[self.step_index]
        await self._converse(prompts.review_prompt(step, diff))
        if self.review is None:
            self.emit(
                Event(
                    kind=EVENT_INFO,
                    text=(
                        "The navigator did not submit a verdict; treating this step "
                        "as still in progress."
                    ),
                )
            )
            return
        self.emit(
            Event(
                kind=EVENT_REVIEW,
                verdict=self.review.verdict,
                text=self.review.feedback,
                step=self.step_index,
            )
        )
        if self.review.verdict != "pass":
            self.persist()
            return

        self.step_index += 1
        if self.step_index >= len(self.plan.steps):
            self.state = STATE_DONE
            self.persist()
            self.emit(Event(kind=EVENT_SESSION_DONE))
            return
        await self._converse(
            prompts.instruct_prompt(self.step_index, self.plan.steps[self.step_index])
        )
        self.persist()
        self.emit(Event(kind=EVENT_STEP_STARTED, step=self.step_index))

    async def skip(self) -> None:
        if self.state != STATE_DRIVE:
            raise RuntimeError("no step in progress")
        skipped = self.step_index
        self._snapshot()
        self.step_index += 1
        if self.step_index >= len(self.plan.steps):
            self.state = STATE_DONE
            self.persist()
            self.emit(Event(kind=EVENT_SESSION_DONE))
            return
        text = (
            prompts.skip_prompt(skipped, self.plan.steps[skipped])
            + "\n\n"
            + prompts.instruct_prompt(self.step_index, self.plan.steps[self.step_index])
        )
        await self._converse(text)
        self.persist()
        self.emit(Event(kind=EVENT_STEP_STARTED, step=self.step_index))

    async def user_message(self, text: str) -> None:
        if self.state == STATE_IDLE:
            raise RuntimeError("session not started")
        await self._converse(text)
        if self.state == STATE_SCAFFOLD:
            if len(self.plan.steps) == 0:
                self.state = STATE_PROPOSE
            else:
                self._snapshot()
                self.state = STATE_DRIVE
                self.emit(Event(kind=EVENT_STEP_STARTED, step=self.step_index))
        self.persist()

    async def summarize(self) -> None:
        if self.state == STATE_IDLE:
            raise RuntimeError("session not started")
        text = await self._converse(prompts.summary_prompt())
        self.persist()
        if text.strip() == "":
            return
        name = f"summary-{self.session_id}.md"
        path = Path(self.dir) / ".nina" / name
        path.parent.mkdir(parents=True, exist_ok=True)
        path.write_text(text + "\n")
        self.emit(Event(kind=EVENT_INFO, text=f"Summary saved to `.nina/{name}`"))

    async def _converse(self, text: str) -> str:
        assert self.session is not None
        self.review = None
        parts: list[str] = []
        async for event in self.session.send(text):
            if isinstance(event, TextDelta):
                parts.append(event.text)
                if self.review is None:
                    self.emit(Event(kind=EVENT_TEXT_DELTA, text=event.text))
            elif isinstance(event, RateLimited):
                self.emit(Event(kind=EVENT_INFO, text=f"Rate limited ({event.rate_limit_type})."))
            elif isinstance(event, TurnComplete):
                break
        return "".join(parts)

    async def _write_file(self, args: dict[str, object]) -> ToolResult:
        if self.profile.dial == 0:
            return ToolResult(
                "Denied by the typing dial policy: at dial level 0 you may not write files "
                "at all. Instruct the learner to write this themselves.",
                is_error=True,
            )
        if self.profile.dial == 1 and self.state != STATE_SCAFFOLD:
            return ToolResult(
                "Denied by the typing dial policy: at dial level 1 you may only write "
                "files while scaffolding the project. Instruct the learner to write this "
                "code themselves.",
                is_error=True,
            )
        path_arg = str(args.get("path", ""))
        content = str(args.get("content", ""))
        try:
            path = self._safe_path(path_arg)
        except ValueError as err:
            return ToolResult(str(err), is_error=True)
        path.parent.mkdir(parents=True, exist_ok=True)
        path.write_text(content)
        self.emit(Event(kind=EVENT_INFO, text=f"Nina wrote `{path_arg}`"))
        return ToolResult(f"wrote {path_arg}")

    async def _set_plan(self, args: dict[str, object]) -> ToolResult:
        if self.state not in (STATE_PROPOSE, STATE_SCAFFOLD):
            return ToolResult(
                "the session already has a plan; revise the remaining steps with "
                "update_plan instead",
                is_error=True,
            )
        try:
            steps = plan_steps_from(args.get("steps"))
        except ValueError as err:
            return ToolResult(f"invalid set_plan input: {err}", is_error=True)
        if not steps:
            return ToolResult("a plan needs at least one step", is_error=True)
        self.plan = Plan(title=str(args.get("title", "")), steps=steps)
        if self.state == STATE_PROPOSE:
            self.state = STATE_SCAFFOLD
        self.emit(Event(kind=EVENT_PLAN_SET, plan=self.plan))
        return ToolResult(f"plan set with {len(steps)} steps")

    async def _update_plan(self, args: dict[str, object]) -> ToolResult:
        if self.state != STATE_DRIVE:
            return ToolResult(
                "update_plan is only available once the session is underway; use "
                "set_plan for the initial plan",
                is_error=True,
            )
        try:
            steps = plan_steps_from(args.get("steps"))
        except ValueError as err:
            return ToolResult(f"invalid update_plan input: {err}", is_error=True)
        if not steps:
            return ToolResult("update_plan needs at least one replacement step", is_error=True)
        self.plan.steps = self.plan.steps[: self.step_index + 1] + steps
        self.persist()
        self.emit(Event(kind=EVENT_PLAN_SET, plan=self.plan))
        return ToolResult(f"plan revised: {len(steps)} steps remain after the current one")

    async def _submit_review(self, args: dict[str, object]) -> ToolResult:
        verdict = str(args.get("verdict", ""))
        feedback = str(args.get("feedback", ""))
        if verdict not in ("pass", "retry"):
            return ToolResult("verdict must be pass or retry", is_error=True)
        self.review = SubmitReviewInput(verdict=verdict, feedback=feedback)
        return ToolResult("verdict recorded")

    async def _run_command(self, args: dict[str, object]) -> ToolResult:
        command = str(args.get("command", "")).strip()
        reason = str(args.get("reason", ""))
        if not command:
            return ToolResult("run_command needs a non-empty command", is_error=True)
        if command not in self._auto_approve:
            future: asyncio.Future[ConfirmAnswer] = asyncio.get_running_loop().create_future()
            self.emit(
                Event(
                    kind=EVENT_CONFIRM,
                    confirm=ConfirmRequest(command=command, reason=reason, reply=future),
                )
            )
            answer = await future
            if answer.approve and answer.always:
                self._auto_approve.add(command)
            if not answer.approve:
                return ToolResult(
                    "The learner declined to run this command. Continue without it, or "
                    "propose an alternative and explain why it is needed."
                )
        result = await runner.run(self.dir, command, COMMAND_TIMEOUT)
        self.emit(Event(kind=EVENT_COMMAND_RUN, text=_command_output_markdown(command, result)))
        return ToolResult(_command_tool_content(result))

    async def _read_file(self, args: dict[str, object]) -> ToolResult:
        path_arg = str(args.get("path", ""))
        try:
            path = self._safe_path(path_arg)
        except ValueError as err:
            return ToolResult(str(err), is_error=True)
        try:
            data = path.read_bytes()
        except OSError as err:
            return ToolResult(str(err), is_error=True)
        if len(data) > MAX_READ_FILE_BYTES:
            return ToolResult(
                data[:MAX_READ_FILE_BYTES].decode(errors="replace") + "\n[file truncated]"
            )
        return ToolResult(data.decode(errors="replace"))

    def _safe_path(self, rel: str) -> Path:
        if not rel or Path(rel).is_absolute():
            raise ValueError(f"path must be relative to the workspace: {rel!r}")
        root = Path(self.dir).resolve()
        path = (root / rel).resolve()
        if path != root and root not in path.parents:
            raise ValueError(f"path escapes the workspace: {rel!r}")
        return path

    def _snapshot(self) -> None:
        ref = snapshot_ref(self.session_id, self.snapshots)
        self.ws.snapshot(ref)
        self.last_ref = ref
        self.snapshots += 1


def _command_output_markdown(command: str, result: Result) -> str:
    body = f"```console\n$ {command}\n"
    if result.output:
        body += result.output + "\n"
    body += "```\n"
    if result.timed_out:
        body += f"*timed out after {int(COMMAND_TIMEOUT)}s and was killed*"
    elif result.exit_code == 0:
        body += "*exit code 0*"
    else:
        body += f"*exit code {result.exit_code}*"
    return body


def _command_tool_content(result: Result) -> str:
    content = f"exit code: {result.exit_code}"
    if result.timed_out:
        content = f"command timed out after {int(COMMAND_TIMEOUT)}s and was killed\n{content}"
    if not result.output:
        return content + "\n(no output)"
    return content + "\n" + result.output


def new_engine(
    ws: Workspace,
    dir: str,
    prof: Profile,
    emit: Callable[[Event], None],
    session_factory: Callable[[str, dict[str, ToolHandler]], AgentSession],
) -> Engine:
    engine = Engine(ws, dir, prof, emit)
    engine.session = session_factory(engine.system_prompt, engine.tool_handlers())
    return engine
